File: platform/lib-java/src/test/java/com/canary/platform/lib/observability/CanaryMetricsTest.java
package com.canary.platform.lib.observability;

import com.canary.platform.lib.XCanaryContext;
import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Tags;
import io.micrometer.core.instrument.Timer;
import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.time.Duration;

import static org.assertj.core.api.Assertions.assertThat;

class CanaryMetricsTest {

    private MeterRegistry registry;
    private CanaryMetrics metrics;

    @BeforeEach
    void setUp() {
        registry = new SimpleMeterRegistry();
        metrics = new CanaryMetrics(registry, "payment");
    }

    @AfterEach
    void tearDown() {
        XCanaryContext.clear();
    }

    @Test
    void recordHttpIncrementsCounterWithExpectedTags() {
        XCanaryContext.set(false);
        metrics.recordHttp("POST /pay", "success", Duration.ofMillis(42));

        Counter c = registry.find("canary_request_total")
                .tags("substrate", "http", "service", "payment", "lane", "stable",
                      "outcome", "success", "target", "POST /pay")
                .counter();
        assertThat(c).isNotNull();
        assertThat(c.count()).isEqualTo(1.0);
    }

    @Test
    void recordHttpRecordsTimerWithExpectedTags() {
        XCanaryContext.set(true);
        metrics.recordHttp("GET /healthz", "success", Duration.ofMillis(7));

        Timer t = registry.find("canary_request_duration_seconds")
                .tags("substrate", "http", "service", "payment", "lane", "canary",
                      "target", "GET /healthz")
                .timer();
        assertThat(t).isNotNull();
        assertThat(t.count()).isEqualTo(1L);
        assertThat(t.totalTime(java.util.concurrent.TimeUnit.MILLISECONDS)).isEqualTo(7.0);
    }

    @Test
    void recordKafkaUsesKafkaSubstrateAndTopicTarget() {
        XCanaryContext.set(true);
        metrics.recordKafka("payments.charged", "success", Duration.ofMillis(100));

        Counter c = registry.find("canary_request_total")
                .tags("substrate", "kafka", "service", "payment", "lane", "canary",
                      "outcome", "success", "target", "payments.charged")
                .counter();
        assertThat(c).isNotNull();
        assertThat(c.count()).isEqualTo(1.0);
    }

    @Test
    void recordRestateUsesRestateSubstrateAndHandlerTarget() {
        XCanaryContext.set(false);
        metrics.recordRestate("CheckoutSagaStable.run", "server_error", Duration.ofMillis(250));

        Counter c = registry.find("canary_request_total")
                .tags("substrate", "restate", "service", "payment", "lane", "stable",
                      "outcome", "server_error", "target", "CheckoutSagaStable.run")
                .counter();
        assertThat(c).isNotNull();
        assertThat(c.count()).isEqualTo(1.0);
    }

    @Test
    void recordShadowMismatchIncrementsByService() {
        metrics.recordShadowMismatch("totalCents");

        Counter c = registry.find("canary_shadow_mismatch_total")
                .tags("service", "payment", "field", "totalCents")
                .counter();
        assertThat(c).isNotNull();
        assertThat(c.count()).isEqualTo(1.0);
    }
}

File: platform/lib-node/src/__tests__/canary-metrics.test.ts
import { Registry } from "prom-client";
import { describe, expect, it } from "vitest";
import { runWithCanary } from "../x-canary-context.js";
import { CanaryMetrics } from "../observability/canary-metrics.js";

describe("CanaryMetrics", () => {
  it("recordHttp increments counter with expected labels", async () => {
    const registry = new Registry();
    const metrics = new CanaryMetrics("payment", registry);

    await runWithCanary(false, async () => {
      metrics.recordHttp("POST /pay", "success", 0.042);
    });

    const c = await registry.getSingleMetric("canary_request_total")?.get();
    expect(c?.values).toContainEqual(
      expect.objectContaining({
        value: 1,
        labels: { substrate: "http", service: "payment", lane: "stable", outcome: "success", target: "POST /pay" },
      }),
    );
  });

  it("recordHttp records histogram with expected labels", async () => {
    const registry = new Registry();
    const metrics = new CanaryMetrics("payment", registry);

    await runWithCanary(true, async () => {
      metrics.recordHttp("GET /healthz", "success", 0.007);
    });

    const h = await registry.getSingleMetric("canary_request_duration_seconds")?.get();
    const sum = h?.values.find((v) =>
      v.metricName === "canary_request_duration_seconds_sum" &&
      v.labels.target === "GET /healthz" &&
      v.labels.lane === "canary",
    );
    expect(sum?.value).toBeCloseTo(0.007);
  });

  it("recordKafka uses kafka substrate and topic target", async () => {
    const registry = new Registry();
    const metrics = new CanaryMetrics("audit", registry);

    await runWithCanary(true, async () => {
      metrics.recordKafka("payments.charged", "success", 0.1);
    });

    const c = await registry.getSingleMetric("canary_request_total")?.get();
    expect(c?.values).toContainEqual(
      expect.objectContaining({
        value: 1,
        labels: { substrate: "kafka", service: "audit", lane: "canary", outcome: "success", target: "payments.charged" },
      }),
    );
  });

  it("recordRestate uses restate substrate and handler target", async () => {
    const registry = new Registry();
    const metrics = new CanaryMetrics("order", registry);

    metrics.recordRestate("CheckoutSagaStable.run", "server_error", 0.25);

    const c = await registry.getSingleMetric("canary_request_total")?.get();
    expect(c?.values).toContainEqual(
      expect.objectContaining({
        value: 1,
        labels: { substrate: "restate", service: "order", lane: "stable", outcome: "server_error", target: "CheckoutSagaStable.run" },
      }),
    );
  });

  it("recordShadowMismatch increments by service+field", async () => {
    const registry = new Registry();
    const metrics = new CanaryMetrics("payment", registry);

    metrics.recordShadowMismatch("totalCents");

    const c = await registry.getSingleMetric("canary_shadow_mismatch_total")?.get();
    expect(c?.values).toContainEqual(
      expect.objectContaining({
        value: 1,
        labels: { service: "payment", field: "totalCents" },
      }),
    );
  });
});
